core/src/main/java/org/apache/iceberg/SnapshotOperations.java

```diff
@@ -0,0 +1,202 @@
+    this.snapshots = Lists.newArrayList(snapshots);
+    this.snapshotsSupplier = snapshotsSupplier;
+    this.refs = Maps.newHashMap(refs);
+
+    this.snapshotsById = indexSnapshotsById(snapshots);
+  }
+
+  List<Snapshot> snapshots() {
+    ensureLoaded();
+
+    return Collections.unmodifiableList(snapshots);
+  }
+
+  Snapshot snapshot(long id) {
+    if (!snapshotsById.containsKey(id)) {
+      ensureLoaded();
+    }
+
+    return snapshotsById.get(id);
+  }
+
+  boolean contains(long id) {
+    if (!snapshotsById.containsKey(id)) {
+      ensureLoaded();
+    }
+
+    return snapshotsById.containsKey(id);
+  }
+
+  void add(Snapshot snapshot) {
+    ValidationException.check(
+        !contains(snapshot.snapshotId()),
+        "Snapshot already exists for id: %s",
+        snapshot.snapshotId());
+
+    snapshots.add(snapshot);
+    snapshotsById.put(snapshot.snapshotId(), snapshot);
+  }
+
+  List<Snapshot> remove(Collection<Long> idsToRemove) {
+    List<Snapshot> retainedSnapshots =
+        Lists.newArrayListWithExpectedSize(snapshots.size() - idsToRemove.size());
+    List<Snapshot> removedSnapshots = Lists.newArrayListWithExpectedSize(idsToRemove.size());
+
+    for (Snapshot snapshot : snapshots) {
+      long snapshotId = snapshot.snapshotId();
+      if (idsToRemove.contains(snapshotId)) {
+        Snapshot removed = snapshotsById.remove(snapshotId);
+        removedSnapshots.add(removed);
+      } else {
+        retainedSnapshots.add(snapshot);
+      }
+    }
+
+    this.snapshots = retainedSnapshots;
+
+    // remove any refs that are no longer valid
+    Set<String> danglingRefs = Sets.newHashSet();
+    for (Map.Entry<String, SnapshotRef> refEntry : refs.entrySet()) {
+      if (!snapshotsById.containsKey(refEntry.getValue().snapshotId())) {
+        danglingRefs.add(refEntry.getKey());
+      }
+    }
+
+    danglingRefs.forEach(this::removeRef);
+
+    return removedSnapshots;
+  }
+
+  Map<String, SnapshotRef> refs() {
+    return Collections.unmodifiableMap(refs);
+  }
+
+  SnapshotRef ref(String name) {
+    return refs.get(name);
+  }
+
+  void addRef(String name, SnapshotRef ref) {
+    refs.put(name, ref);
+  }
+
+  SnapshotRef removeRef(String name) {
+    return refs.remove(name);
+  }
+
+  private void ensureLoaded() {
+    if (snapshotsSupplier != null) {
+      snapshots.addAll(snapshotsSupplier.get());
+      snapshotsSupplier = null;
+
+      this.snapshotsById = indexSnapshotsById(snapshots);
+    }
+  }
+
+  void validate(long currentSnapshotId, long lastSequenceNumber) {
+    validateSnapshots(lastSequenceNumber);
+    validateRefs(currentSnapshotId);
+  }
+
+  private void validateSnapshots(long lastSequenceNumber) {
+    for (Snapshot snap : snapshots) {
+      ValidationException.check(
+          snap.sequenceNumber() <= lastSequenceNumber,
+          "Invalid snapshot with sequence number %s greater than last sequence number %s",
+          snap.sequenceNumber(),
+          lastSequenceNumber);
```

Comment:
I think there's a potential problem here. The supplier can return snapshots that have been committed but aren't in the local node's loaded metadata file. If that happens, this check might fail.

Maybe we should filter the snapshots that are loaded from the supplier based on the last sequence number from the loaded metadata file.